### DISCUSSION TITLE
Cache pulled docker images (already built and published)

### DISCUSSION THREAD

**Discussion by guillermoprados** [community/community#26330] [Actions]
<p>I have a workflow that needs to download a couple of heavy images already built from dockerhub registry. I would like to cache them instead download them everytime I run a workflow, but I can’t figure out how to do it (If there is a way)</p>
<p>I have been reading how to use <a href="https://help.github.com/en/actions/configuring-and-managing-workflows/caching-dependencies-to-speed-up-workflows" rel="nofollow noopener">actions cache</a> , and also some suggestions <a href="https://github.community/t/cache-a-docker-image-built-in-workflow/16260">here</a> and <a href="https://github.community/t/cache-a-docker-image-built-in-workflow/16260">here</a> , but couldn’t find a way to actually cache the image there, because I cannot find the path where docker pulls  the images</p>
<p>I’ve been playing with this action too:</p>
<p><a href="https://github.community/t/cache-docker-images/17521">https://github.community/t5/GitHub-Actions/Cache-docker-images/td-p/44978</a><br>
without any luck… I was able to build my image using this, but it get’s rebuild every time I run my workflows… and using this, I now have another image getting downloaded in order to run the action… so it’s even worse.</p>
<p>Spent an entire day trying to find the folder where docker saves the images in the runner without any luck… it’s driving me crazy.</p>
<p>My other doubt is… why is nobody asking about this? Isn’t this a common practice? we have almost 5GB of available cache, and I think that storing any image there will certainly speed the workflow time… but maybe I’m misundertanding something?</p>

**Comment by lucianposton** [comment]
<p>It wasn’t clear how the docker image is downloaded during your workflow.</p>
<p>If you’re downloading it via docker CLI, the following may do what you need.</p>
<pre><code>test_docker_image:
        runs-on: ubuntu-latest
        steps:
            - uses: actions/checkout@v2

            - run: mkdir -p ~/image-cache

            - id: image-cache
              uses: actions/cache@v1
              with:
                  path: ~/image-cache
                  # Adjust key to meet your cache time requirements e.g.
                  # ${{ hashFiles(*) }} can be useful here to invalidate
                  # cache on file changes
                  key: image-cache-${{ runner.os }}

            - if: steps.image-cache.outputs.cache-hit != 'true'
              run: |
                  docker pull alpine
                  docker save -o ~/image-cache/alpine.tar alpine

            - if: steps.image-cache.outputs.cache-hit == 'true'
              run: docker load -i ~/image-cache/alpine.tar

            - name: 'Run tests on image (or whatever)'
              uses: whatever/structure-tests-action@master
              with:
                  image: alpine
                  configFile: 'test/structure-tests.yaml'
</code></pre>

**Comment by BrightRan** [comment]
<p><a class="mention" href="/guillermoprados">@guillermoprados</a> ,</p>
<p>Currently, GitHub’s cache action does not support to cache the builds of Docker images.</p>
<p>I noticed the below two tickets which had reported the same question in the repository of cache action:</p>
<p><a href="https://github.com/actions/cache/issues/81" rel="nofollow noopener">https://github.com/actions/cache/issues/81</a></p>
<p><a href="https://github.com/actions/cache/issues/31" rel="nofollow noopener">https://github.com/actions/cache/issues/31</a></p>
<p>You can follow the two tickets and add your comments on them, that will allow you to directly interact with the appropriate engineering team.</p>

**Comment by lucianposton** [comment, marked as answer]
<p>&gt; GitHub’s cache action does not support to cache the builds of Docker images.</p>
<p>It doesn’t automatically cache docker images for you, no (I believe that’s what you’re saying).</p>
<p>However, github’s cache action does work with docker image tar files, which are created/restored using <code>docker save</code> and <code>docker load</code>. I have this working with built images (as opposed to pulled images from docker hub), but I suspect it will also work with pulled images.</p>

**Comment by guillermoprados** [comment]
<p>thanks <a class="mention" href="/brightran">@brightran</a> I didn’t see those threads… I need to become a better googler :D. At least I know now I’m not crazy for thinking on it.</p>
<p>On the bright side, I have to recognize the pulling time for the docker hub registry is really impresive.</p>

**Comment by guillermoprados** [comment]
<p><a class="mention" href="/lucianposton">@lucianposton</a> that’s great! I didn’t know about that command.</p>
<p>I’ll test it later to see how much time and space it actually consumes, and I’ll evaluate later if I want to use this apprach or if I`ll continue just pulling from the docker hub registry.</p>
<p>Anyway, since there doesn’t seems to be another solution, I’ll mark this answer as the correct one :) </p>

**Comment by guillermoprados** [comment]
<p>you’re awesome!! thanks man :) </p>

**Comment by OskarPersson** [comment]
<p>Have you been able to get this working in Windows? Using the same configuration that you describe I get an error during <strong>docker save</strong> :</p>
<pre><code>failed to save image: invalid output path: directory "~\\image-cache" does not exist
</code></pre>

**Comment by Kurt-von-Laven** [comment]
<p>We ran into a similar issue and built a dirt-simple <a href="https://github.com/marketplace/actions/docker-cache" rel="noopener nofollow ugc">docker-cache</a> action that uses the <code>docker load</code> + <code>docker save</code> technique described above. I haven’t tested it on Windows, but I see no reason it wouldn’t work cross-platform unless Bash is at a different path on GitHub’s Windows images, which would be trivial for me to fix. For even better performance, you could alternatively upload the Docker images you are pulling to the GitHub Container Registry and simply pull them from there, although that probably doesn’t help if a dependency is doing the pulling.</p>
